### PR TITLE
BitMEX Live Market Data Feed

### PR DESCRIPTION
**Pull request by Ameobea**
This PR adds support for a live market data feed for [BitMEX](https://www.bitmex.com/).  Its structure is largely based off of the Gemini live market data feed.

### CODE REVIEW

File: src/exchanges/bitmex/BitmexMarketFeed.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    protected onOpen(): void {
+        // Nothing for now
+    }
+
+    protected handleMessage(rawMsg: string): void {
+        const msg: BitmexMessage = JSON.parse(rawMsg) as BitmexMessage;
+
+        if (msg.error) {
```
Comment (CjS77) [inline src/exchanges/bitmex/BitmexMarketFeed.ts]: The way we handle error messages in the feed (as opposed to network errors or runtime exceptions) is to pass an `ErrorMessage` instance down the feed. This allows downstream components to handle the message as they see fit.

File: src/exchanges/bitmex/BitmexMarketFeed.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        } else if (msg.success !== undefined) {
+            // subscription response
+            const subscriptionResMSg: SubscriptionResponseMessage = msg as SubscriptionResponseMessage;
+            this.handleSubscriptionSuccess(subscriptionResMSg);
+        } else if (msg.info) {
+            // welcome message
+            this.logger.log('debug', 'Received welcome message from BitMEX WS feed.');
+        } else {
+            // unhandled/unexpected message
+            const err = new Error(`An unhandled message was received over the websocket connection: ${JSON.stringify(rawMsg)}`);
```
Comment (CjS77) [inline src/exchanges/bitmex/BitmexMarketFeed.ts]: Just wrap this message in an `UnknownMessage` and pass it along

File: src/exchanges/bitmex/BitmexMarketFeed.ts [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    private handleSnapshot(snapshot: OrderbookSnapshotMessage) {
+        // (re)initialize our order id map
+        const newIdMap = snapshot.data.reduce((acc, { id, price }) => ({...acc, [id]: price }), {});
+        this.orderIdMap = newIdMap;
+
+        const mapLevelUpdates: (date: PriceData) => PriceLevelWithOrders =
+            ({ id, price, size, side }) => PriceLevelFactory(price, size, side.toLowerCase());
+
+        const asks: PriceLevelWithOrders[] = snapshot.data
+            .filter(R.propEq('side', 'Sell'))
```
Comment (CjS77) [inline src/exchanges/bitmex/BitmexMarketFeed.ts]: We try not to add libraries unnecessarily. This is the only place you use ramda and can easily be replaced by the native `Array.filter` method.

**Comment by Ameobea** [comment]
@CjS77 thanks for the speedy review; I've made the changes you mentioned and fixed the linter rules so the project now builds without issue.

Here's a pared-down diff of the changes since you last reviewed: https://github.com/coinbase/gdax-tt/compare/2c7cfe4...9382f89

**Comment by CjS77** [comment]
I have experienced issues in the past with the hardcoded message types in the message interfaces when writing more dynamic code (which is why they weren't until now).
Hopefully, they don't crop up again, because I prefer it this way too.

**Pull request merged**